src/table_properties.rs

```diff
@@ -0,0 +1,272 @@
+use std::ffi::{c_char, c_void, CStr};
+use std::marker::PhantomData;
+use std::mem;
+use std::slice;
+
+use libc::{c_int, size_t};
+
+use crate::{ffi, Options};
+
+/// Extension trait for [`Options`] to register table properties collectors
+pub trait TablePropertiesExt {
+    fn add_table_properties_collector_factory<F, K, V>(&mut self, factory: F)
+    where
+        F: TablePropertiesCollectorFactory<K, V> + Send + 'static,
+        K: AsRef<CStr>,
+        V: AsRef<CStr>;
+}
+
+impl TablePropertiesExt for Options {
+    fn add_table_properties_collector_factory<F, K, V>(&mut self, factory: F)
+    where
+        F: TablePropertiesCollectorFactory<K, V> + Send + 'static,
+        K: AsRef<CStr>,
+        V: AsRef<CStr>,
+    {
+        unsafe {
+            let factory_ptr = Box::into_raw(Box::new(factory)) as *mut c_void;
+
+            // Takes ownership of the collector factory; the Rust callback wrapper will be
+            // dropped via the destructor callback
+            ffi::rocksdb_options_add_table_properties_collector_factory(
+                self.inner,
+                factory_ptr,
+                Some(TablePropertiesCollectorFactoryCallback::<F, K, V>::destructor),
+                Some(TablePropertiesCollectorFactoryCallback::<F, K, V>::name),
+                Some(TablePropertiesCollectorFactoryCallback::<F, K, V>::create_collector),
+            );
+        }
+    }
+}
+
+#[repr(C)]
+pub enum EntryType {
+    EntryPut,
+    EntryDelete,
+    EntrySingleDelete,
+    EntryMerge,
+    EntryRangeDeletion,
+    EntryBlobIndex,
+    EntryDeleteWithTimestamp,
+    EntryWideColumnEntity,
+    EntryTimedPut,
+    EntryOther,
+}
+
+pub struct TablePropertiesCollectorContext {
+    pub column_family_id: u32,
+    pub level_at_creation: i32,
+    pub num_levels: i32,
+    pub last_level_inclusive_max_seqno_threshold: u64,
+}
+
+/// Table properties collector factory trait
+pub trait TablePropertiesCollectorFactory<K, V>
+where
+    K: AsRef<CStr>,
+    V: AsRef<CStr>,
+{
+    type Collector: TablePropertiesCollector<K, V>;
+
+    /// Create a new table properties collector
+    fn create(&mut self, context: TablePropertiesCollectorContext) -> Self::Collector;
+
+    /// Name of the collector factory to use for logging
+    fn name(&self) -> &CStr;
+}
+
+/// Table properties collector trait
+pub trait TablePropertiesCollector<K, V>
+where
+    K: AsRef<CStr>,
+    V: AsRef<CStr>,
```

Comment:
I was hoping to make this something more like `Borrow<CStrLike>` but hit multiple issues making the generics and ownership/references work. As it stands, this allows the implementation to return an `Iter` or a collection like `Vec` which implements `IntoIter` over a bag of `CStr`s or `CString` references. I think I can easily make this take `&str` as well but then the error handling of Rust to C string conversion is buried in the callback shim, and I really don't like that.